src/terrain/KM_Terrain.pas

```diff
@@ -555,20 +564,60 @@ function TKMTerrain.ScriptTryObjectSet(X, Y: Integer; aObject: Byte): Boolean;
   //Is this object part of a wine/corn field?
   or TileIsWineField(KMPoint(X, Y)) or TileIsCornField(KMPoint(X, Y))
   //Is there a house/site near this object?
-  or HousesNearObject then
+  or HousesNearObject
+  //Is this object allowed to be placed - like in KaM Editor?
+  or not AllowableObject then
   begin
     Result := False;
     Exit;
   end;
 
+
+
   //Did block diagonal property change? (hence xor) UpdateWalkConnect needs to know
   DiagonalChanged := MapElem[Land[Y,X].Obj].DiagonalBlocked xor MapElem[aObject].DiagonalBlocked;
 
   //Apply change
+  //UpdatePassability and UpdateWalkConnect is used in SetField so that we only use it in trees and other objects
   Land[Y, X].Obj := aObject;
-  UpdatePassability(KMRect(X, Y, X, Y)); //When using KMRect map bounds are checked by UpdatePassability
-  UpdateWalkConnect([wcWalk, wcRoad, wcWork], KMRectGrowTopLeft(KMRect(X, Y, X, Y)), DiagonalChanged);
-  Result := True;
+  case aObject of
+
+    55..58: // Wine in different stages
+      if CanAddField(X, Y, ft_Wine) then
+      begin
+         SetField(KMPoint(X, Y), -1, ft_Wine);
+         Result := True;
+      end
+      else
+        Result := False;
+    59..63: // Corn in different stages
+      if CanAddField(X, Y, ft_Corn) then
+      begin
+        SetField(KMPoint(X, Y), -1, ft_Corn);
+        Result := True;
+      end
+      else
+        Result := False;

```

Comment:
When we return False, does this imply that the whole operation failed? And if it has failed, then maybe it should not have the "Land[Y, X].Obj := aObject;" set.